File: DataEngineering/02_Data_Modeling/projects/golang/04_scd_type_2_implementer/main.go
package main

import (
	"fmt"
	"log"
	"time"

	"gorm.io/driver/postgres"
	"gorm.io/gorm"
	"gorm.io/gorm/logger"
)

// DimCustomer represents an SCD Type 2 dimension table
type DimCustomer struct {
	CustomerSK int       `gorm:"primaryKey;autoIncrement;column:customer_sk"`
	CustomerID int       `gorm:"not null;column:customer_id"`
	City       string    `gorm:"not null"`
	StartDate  time.Time `gorm:"not null"`
	EndDate    time.Time `gorm:"not null"`
	IsCurrent  bool      `gorm:"not null;default:true"`
}

// processSCD2Update handles inserting or updating an SCD Type 2 record transactionally.
func processSCD2Update(db *gorm.DB, customerID int, newCity string, effectiveDate time.Time) error {
	return db.Transaction(func(tx *gorm.DB) error {
		// 1. Find the current active record
		var currentRecord DimCustomer
		err := tx.Where("customer_id = ? AND is_current = ?", customerID, true).First(&currentRecord).Error

		if err == nil { // Record exists, we need to expire it
			currentRecord.EndDate = effectiveDate
			currentRecord.IsCurrent = false
			if err := tx.Save(&currentRecord).Error; err != nil {
				return err
			}
		} else if err != gorm.ErrRecordNotFound {
			return err
		}

		// 2. Insert the new active record
		maxDate := time.Date(9999, 12, 31, 0, 0, 0, 0, time.UTC)
		newRecord := DimCustomer{
			CustomerID: customerID,
			City:       newCity,
			StartDate:  effectiveDate,
			EndDate:    maxDate,
			IsCurrent:  true,
		}
		if err := tx.Create(&newRecord).Error; err != nil {
			return err
		}

		return nil
	})
}

func main() {
	dsn := "host=localhost user=postgres password=postgres dbname=data_warehouse port=5432 sslmode=disable"
	db, err := gorm.Open(postgres.Open(dsn), &gorm.Config{
		Logger: logger.Default.LogMode(logger.Info),
	})
	if err != nil {
		log.Printf("Failed to connect to database: %v", err)
		log.Println("Skipping actual execution...")
		return
	}

	// 1. Setup SCD Type 2 Table
	err = db.AutoMigrate(&DimCustomer{})
	if err != nil {
		log.Fatalf("Migration failed: %v", err)
	}
	fmt.Println("SCD Type 2 Table successfully migrated!")

	// 2. Insert initial record (Customer lives in New York)
	t1 := time.Date(2023, 1, 1, 0, 0, 0, 0, time.UTC)
	if err := processSCD2Update(db, 1, "New York", t1); err != nil {
		log.Fatalf("Failed to process initial insert: %v", err)
	}

	// 3. Process Update (Customer moves to Los Angeles)
	t2 := time.Date(2023, 10, 1, 0, 0, 0, 0, time.UTC)
	if err := processSCD2Update(db, 1, "Los Angeles", t2); err != nil {
		log.Fatalf("Failed to process SCD2 update: %v", err)
	}

	// 4. View history
	var records []DimCustomer
	db.Order("customer_sk ASC").Find(&records)

	fmt.Println("--- SCD Type 2 Customer History ---")
	for _, r := range records {
		fmt.Printf("SK: %d | ID: %d | City: %-12s | Current: %-5v | %s to %s\n",
			r.CustomerSK, r.CustomerID, r.City, r.IsCurrent,
			r.StartDate.Format("2006-01-02"), r.EndDate.Format("2006-01-02"))
	}
}

File: DataEngineering/02_Data_Modeling/projects/python/04_scd_type_2_implementer/main.py
import logging
from datetime import datetime, date
from sqlalchemy import create_engine, Column, Integer, String, Date, Boolean
from sqlalchemy.orm import declarative_base, sessionmaker

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

Base = declarative_base()

class DimCustomer(Base):
    """
    SCD Type 2 Dimension Table.
    Tracks historical changes using start_date, end_date, and is_current flag.
    """
    __tablename__ = 'dim_customer'
    
    surrogate_key = Column(Integer, primary_key=True, autoincrement=True)
    customer_id = Column(String(50), nullable=False) # Natural/Business Key
    name = Column(String(100), nullable=False)
    city = Column(String(100), nullable=False)
    
    # SCD Type 2 tracking columns
    start_date = Column(Date, nullable=False)
    end_date = Column(Date, nullable=False)
    is_current = Column(Boolean, nullable=False, default=True)

def apply_scd2_update(session, customer_id: str, new_name: str, new_city: str, effective_date: date):
    """
    Applies an SCD Type 2 update by closing out the current record
    and inserting a new current record.
    """
    # 1. Find the current active record
    current_record = session.query(DimCustomer).filter_by(
        customer_id=customer_id, 
        is_current=True
    ).first()

    if current_record:
        # Step A: Close the old record
        current_record.end_date = effective_date
        current_record.is_current = False
        session.add(current_record)

    # Step B: Insert the new active record
    new_record = DimCustomer(
        customer_id=customer_id,
        name=new_name,
        city=new_city,
        start_date=effective_date,
        end_date=date(9999, 12, 31),
        is_current=True
    )
    session.add(new_record)

def main():
    DATABASE_URL = "sqlite:///:memory:"
    engine = create_engine(DATABASE_URL, echo=False)
    Base.metadata.create_all(engine)
    
    Session = sessionmaker(bind=engine)
    session = Session()

    try:
        # 1. Initial State: Customer lives in Chicago
        initial_date = date(2023, 1, 1)
        apply_scd2_update(session, 'C1', 'Bob Jones', 'Chicago', initial_date)
        session.commit()
        
        logger.info("--- Initial State ---")
        for rec in session.query(DimCustomer).all():
            logger.info(f"SK: {rec.surrogate_key} | ID: {rec.customer_id} | City: {rec.city} | Current: {rec.is_current} | {rec.start_date} to {rec.end_date}")
        
        # 2. Update (SCD Type 2 Event): Customer moves to Seattle
        update_date = date(2023, 6, 15)
        apply_scd2_update(session, 'C1', 'Bob Jones', 'Seattle', update_date)
        session.commit()
        
        logger.info("\n--- After SCD Type 2 Update (Moved to Seattle) ---")
        for rec in session.query(DimCustomer).all():
            logger.info(f"SK: {rec.surrogate_key} | ID: {rec.customer_id} | City: {rec.city} | Current: {rec.is_current} | {rec.start_date} to {rec.end_date}")

    except Exception as e:
        session.rollback()
        logger.error(f"Failed to process SCD2 event: {e}")
    finally:
        session.close()

if __name__ == '__main__':
    main()
